src/snowflake/snowpark/modin/plugin/extensions/utils.py

```diff
@@ -72,3 +91,495 @@ def add_cache_result_docstring(func: Fn) -> Fn:
+        result_agg_func = func_dict
+    elif is_list_like(agg_func):
+        # When the input func is in list like format like [min, max, np.sum], perform a processing to the
+        # aggregation function to convert it to string representation if possible.
+        result_agg_func = try_convert_builtin_func_to_str(agg_func, obj)
+    else:
+        result_agg_func = agg_func
+
+    if not allow_duplication:
+        # if allow_duplication is False, check is there duplication in the function names, which
+        # are used as the row label for the aggregation result in dataframe/series aggregation, and
+        # not allowed in pandas.
+        found_duplication = False
+        if is_dict_like(result_agg_func):
+            for agg_func in result_agg_func.values():
+                if is_list_like(agg_func):
+                    agg_func_names = [get_pandas_aggr_func_name(fn) for fn in agg_func]
+                    found_duplication = len(agg_func_names) > len(set(agg_func_names))
+                    break
+        elif is_list_like(result_agg_func):
+            agg_func_names = [get_pandas_aggr_func_name(fn) for fn in result_agg_func]
+            found_duplication = len(agg_func_names) > len(set(agg_func_names))
+
+        if found_duplication:
+            raise SpecificationError("Function names must be unique!")
+
+    return result_agg_func
+
+
+def get_as_shape_compatible_dataframe_or_series(
+    other: pd.DataFrame | pd.Series | Callable | AnyArrayLike | Scalar,
+    reference_df: pd.DataFrame,
+    shape_mismatch_message: None
+    | (str) = "Array conditional must be same shape as self",
+) -> pd.DataFrame | pd.Series:
+    """
+    Get the "other" type as a shape compatible dataframe or series using the reference_df as a reference for
+    compatible shape and construction.  If there is no shape on the other type then wrap as a numpy array.
+
+    Parameters
+    ----------
+        other : Other type which could be array like
+        reference_df : Reference dataframe or series
+
+    Returns
+    -------
+        Dataframe or series that contains same values as other
+    """
+    if not hasattr(other, "shape"):
+        # If an array type is provided that doesn't have a shape, then wrap it so it has a shape.
+        # For example, if other=[1,2,3] then np.asanyarray will wrap as a numpy array with correct shape,
+        # ie, np.anyarray(other).shape=(3,) in this case.
+        other = np.asanyarray(other)
+
+    if len(other.shape) == 0 or other.shape != reference_df.shape:
+        raise ValueError(shape_mismatch_message)
+
+    if "columns" in reference_df:
+        other = reference_df.__constructor__(
+            other, index=reference_df.index, columns=reference_df.columns
+        )
+    else:
+        other = reference_df.__constructor__(other, index=reference_df.index)
+
+    return other
+
+
+def ensure_index(
+    index_like: Axes | pd.Index | pd.Series, copy: bool = False
+) -> pd.Index | pandas.MultiIndex:
```

Comment:
I assume those are copied from the deleted utils file with functions that doesn't exists in the modin utils, right?

Reply:
Yes, that's correct.